### PR TITLE
Feat: Use thread unique workspace, add knowledge ingest

### PR DESCRIPTION
**Pull request by StrongMonkey**
This pr adds the following features/changes/improvements:

1. Add new knowledge ingestion workflow. It ingests file using threadID as datasetID after file is uploaded. It also packages latest knowledge binary releases into the app. It also adds new `GPTSCRIPT_THREAD_ID` to gptscript run so that it can be used by knowledge tool to know which dataset it needs to query from.
2. It changes thread to have its unique workspace. For knowledge files, it will be stored in `knowledge` subdir inside its own workspace.
3. Changed so that when you go to chat, if there is no thread, it will automatically create a thread with tildy. Once you have threads, it will go to the latest thread you interacted with.
4. Make so that loading status is moved to message bar/button instead of whole screen.
5. Chatting with new assistant will by default create a thread.

### CODE REVIEW

File: actions/knowledge/knowledge.ts
Comment (StrongMonkey) [inline actions/knowledge/knowledge.ts:8]: This is where we run knowledge ingest command on server side to ingest file on fly.

File: contexts/script.tsx
Comment (njhale) [inline contexts/script.tsx:184]: nit:

```suggestion
if (!enableThread || thread || threadInitialized.current) {
  return;
}

threadInitialized.current = true;

const createAndSetThread = async () => {
  try {
    const threads = await getThreads();

    if ((initialScript && initialScriptId) || threads.length === 0) {
      const newThread = await createThread(
        script,
        scriptDisplayName ?? defaultScriptName,
        scriptId
      );
      const threadId = newThread?.meta?.id;
      setThread(threadId);
      setThreads(await getThreads());
      setSelectedThreadId(threadId);
      setWorkspace(newThread.meta.workspace);
    } else if (threads.length > 0) {
      const latestThread = threads[0];
      setThreads(threads);
      setThread(latestThread.meta.id);
      setSelectedThreadId(latestThread.meta.id);
      setScriptId(latestThread.meta.scriptId);
    }
  } catch (e) {
    threadInitialized.current = false;
    console.error(e);
  }
};

createAndSetThread();
```

File: electron/build.mjs
Comment (njhale) [inline electron/build.mjs:78]: We should probably add this directory to the `.gitignore`.

File: package.json
Comment (njhale) [inline package.json:11]: I don't think we need this since we're already including the `bin` directory with the `extraResources` field on the config we're passing to electron builder.

File: scripts/install-binary.mjs
Comment (njhale) [inline scripts/install-binary.mjs:144]: `.mjs` files are ES modules, so you should be able to call `await` from the top-level of the script without an IIFE

```suggestion

  await needToInstall();
  if (process.env.KNOWLEDGE_SKIP_INSTALL_BINARY === 'true') {
    console.info('Skipping binary download');
    process.exit(0);
  }

  console.log(`Downloading and extracting knowledge binary from ${url}...`);
  try {
    await downloadAndExtract(url, outputDir);
  } catch (error) {
    console.error('Error downloading and extracting:', error);
  }

```

**Review by njhale — CHANGES_REQUESTED** [review]
Looking great 👍 

I left a few comments, but besides that I'd like to check to make sure the new binaries don't interfere with notarizing before this goes in. I'll circle back with the results shortly.

File: actions/threads.tsx
Comment (njhale) [inline actions/threads.tsx:35]: The first time I started the app, I got this error:

```
Failed to read threads directory Error: ENOENT: no such file or directory, scandir '/Users/nick/Library/Application Support/acorn/Acorn/threads'
    at async readdir (node:internal/fs/promises:948:18)
    at async m (/Users/nick/projects/gptscript-ai/desktop/electron-dist/mac-arm64/Acorn.app/Contents/Resources/app.asar/.next/server/chunks/500.js:6:22272)
    at async /Users/nick/projects/gptscript-ai/desktop/electron-dist/mac-arm64/Acorn.app/Contents/Resources/app.asar/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:16:408
    at async rP (/Users/nick/projects/gptscript-ai/desktop/electron-dist/mac-arm64/Acorn.app/Contents/Resources/app.asar/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:15:8043)
    at async r5 (/Users/nick/projects/gptscript-ai/desktop/electron-dist/mac-arm64/Acorn.app/Contents/Resources/app.asar/node_modules/next/dist/compiled/next-server/app-page.runtime.prod.js:18:1139)
    at async doRender (/Users/nick/projects/gptscript-ai/desktop/electron-dist/mac-arm64/Acorn.app/Contents/Resources/app.asar/node_modules/next/dist/server/base-server.js:1433:30)
    at async cacheEntry.responseCache.get.routeKind (/Users/nick/projects/gptscript-ai/desktop/electron-dist/mac-arm64/Acorn.app/Contents/Resources/app.asar/node_modules/next/dist/server/base-server.js:1594:28)
    at async NextNodeServer.renderToResponseWithComponentsImpl (/Users/nick/projects/gptscript-ai/desktop/electron-dist/mac-arm64/Acorn.app/Contents/Resources/app.asar/node_modules/next/dist/server/base-server.js:1502:28)
    at async NextNodeServer.renderPageComponent (/Users/nick/projects/gptscript-ai/desktop/electron-dist/mac-arm64/Acorn.app/Contents/Resources/app.asar/node_modules/next/dist/server/base-server.js:1919:24)
    at async NextNodeServer.renderToResponseImpl (/Users/nick/projects/gptscript-ai/desktop/electron-dist/mac-arm64/Acorn.app/Contents/Resources/app.asar/node_modules/next/dist/server/base-server.js:1957:32) {
  errno: -2,
  code: 'ENOENT',
  syscall: 'scandir',
  path: '/Users/nick/Library/Application Support/acorn/Acorn/threads'
}
```

It's not really a big deal since it didn't impact usability and I didn't get it on the second run, but does pollute the logs a bit. 

It's probably worth checking the exception for `ENOENT` here to avoid logging the error.

**Review by njhale — CHANGES_REQUESTED** [review]

File: actions/threads.tsx
Comment (StrongMonkey) [reply actions/threads.tsx:35]: Yeah good call. I will fix that

File: components/script/chatBar/commands.tsx
Comment (StrongMonkey) [reply components/script/chatBar/commands.tsx:38]: i will probably leave it here. It is probably going to be changed later

File: actions/threads.tsx
Comment (thedadams) [inline actions/threads.tsx:129]: As I understand it, it was a requirement to have the LLM name the thread based on the first message. Has that requirement changed?

File: components/script.tsx
Comment (thedadams) [inline components/script.tsx:1]: 😍

File: server/app.mjs
Comment (thedadams) [inline server/app.mjs:144]: This shouldn't be necessary. Add env vars here should append to the existing environment.

**Review by thedadams — APPROVED** [review]

File: components/script/chatBar/commands.tsx
Comment (iwilltry42) [reply components/script/chatBar/commands.tsx:38]: The knowledge PR is merged and a tag `v0.4.10-gateway.1` released :+1:

**Review by iwilltry42 — COMMENTED** [review]
LGTM, but since I'm new to the code base, I'd wait for @njhale's and @thedadams' final review :)

File: actions/threads.tsx
Comment (StrongMonkey) [reply actions/threads.tsx:129]: Hmm, I changed it because the thread will always be created at first and it is kind of weird that the name will be changed later after user type the message. But yeah I will confirm with craig.

File: server/app.mjs
Comment (StrongMonkey) [reply server/app.mjs:144]: I looked at the code and it doesn't seem so. https://github.com/gptscript-ai/node-gptscript/blob/main/src/gptscript.ts#L88

File: actions/threads.tsx
Comment (StrongMonkey) [reply actions/threads.tsx:129]: will change this, thanks for pointting out

**Review by njhale — APPROVED** [review]

File: server/app.mjs
Comment (thedadams) [reply server/app.mjs:144]: It happens elsewhere.

**Pull request merged**
